packages/vm/src/traits.rs

```diff
@@ -27,7 +27,35 @@ impl<S: Storage, A: Api, Q: Querier> Extern<S, A, Q> {
 }
 
 #[cfg(feature = "iterator")]
-pub type StorageIteratorItem = FfiResult<(KV, u64)>;
+pub type NextItem = (Option<KV>, u64);
+
+#[cfg(feature = "iterator")]
+pub trait StorageIterator {
+    fn next(&mut self) -> FfiResult<NextItem>;
+
+    /// Collects all elements, ignoring gas costs
+    fn elements(mut self) -> FfiResult<Vec<KV>>
```

Comment:
Instead of implementing this method, you could:
```rust
impl<I: StorageIterator> Iterator for I {
  type Item = Option<KV>;

  fn next(&mut self) -> Self::Item {
    StorageIterator::next(self).0
  }
}
```

and then you get all the iterator methods on types that implement `StorageIterator` for free, so you wouldn't even really need most of the other changes in this PR :)

Reply:
Ok so this is not a thing because:
>implementing a foreign trait is only possible if at least one of the types for which is it implemented is local
>only traits defined in the current crate can be implemented for a type parameter

In any case the "correct" implementation would have been:

```rust
#[cfg(feature = "iterator")]
impl<I: StorageIterator> Iterator for I {
    type Item = Option<FfiResult<KV>>;

    fn next(&mut self) -> Self::Item {
        StorageIterator::next(self).map(|(kv, _gas)| kv).transpose()
    }
}
```

Reply:
I don't like this because there is no natural conversion from `StorageIterator` to `Iterator` possible. Having a dedicated function ensures we can communicate what information are stripped off.

Maybe we can have `fn elements` returning an iterator though

Reply:
But there is no need for more than the absolute minimal solution as this is more a test helper than a real API of `StorageIterator`

Reply:
we could have an `IntoIterator` implementation, or even just a `into_iter` method..?
then you'd be able to get a usual iterator by just calling that method, with the caveat that you won't be getting the gas cost.

Reply:
>But there is no need for more than the absolute minimal solution as this is more a test helper than a real API of StorageIterator

That's a good point... yeah, feel free to just leave it then. If you want to completely hide it, you could even add a `#[cfg(test)]` on it.

Reply:
>  If you want to completely hide it, you could even add a `#[cfg(test)]` on it.

Good idea. I just tried that but our integration tests are not guarded with `#[cfg(test)]`.

I leave it like that for now, keeping a custom `fn into_elements_iter` as a potential future improvements. But this would require at least one more type and one `impl Iterator` for that type.

Reply:
Yeah, i would consider adding another type just for that method to be unnecessary complexity :P
As you said, let's just leave it as it for now :)